cpp/log/frontend.cc

```diff
@@ -111,95 +105,40 @@ std::string Frontend::SubmitResultString(SubmitResult result) {
 }
 
 // static
-SubmitResult Frontend::GetSubmitError(
-    CertSubmissionHandler::SubmitResult result) {
+Status Frontend::GetSubmitError(CertSubmissionHandler::SubmitResult result) {
   CHECK_NE(result, CertSubmissionHandler::OK);
 
-  SubmitResult submit_result;
   switch (result) {
     case CertSubmissionHandler::EMPTY_SUBMISSION:
+      return util::Status(util::error::INVALID_ARGUMENT, "Empty submission.");
     case CertSubmissionHandler::INVALID_PEM_ENCODED_CHAIN:
-      submit_result = BAD_PEM_FORMAT;
-      break;
+      return util::Status(util::error::INVALID_ARGUMENT,
+                          "Invalid PEM encoded chain.");
     case CertSubmissionHandler::SUBMISSION_TOO_LONG:
-      submit_result = SUBMISSION_TOO_LONG;
-      break;
+      return util::Status(util::error::INVALID_ARGUMENT,
+                          "Submission too long.");
     case CertSubmissionHandler::INVALID_CERTIFICATE_CHAIN:
+      return util::Status(util::error::FAILED_PRECONDITION,
+                          "Invalid certificate chain.");
     case CertSubmissionHandler::UNKNOWN_ROOT:
-      submit_result = CERTIFICATE_VERIFY_ERROR;
-      break;
+      return util::Status(util::error::FAILED_PRECONDITION, "Unknown root.");
     case CertSubmissionHandler::PRECERT_CHAIN_NOT_WELL_FORMED:
-      submit_result = PRECERT_CHAIN_NOT_WELL_FORMED;
-      break;
+      return util::Status(util::error::FAILED_PRECONDITION,
```

Comment:
Same here?

For ```UNKNOWN_ROOT```, I'd guess that "adding the root" is something that could be done (the request might otherwise be valid, but not acceptable in the current state of the server), so that's ok.